rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -1939,33 +1939,33 @@ bool spu_thread::process_mfc_cmd()
 					}
 				}
 			}
-			else if (auto& data = vm::_ref<decltype(rdata)>(addr); rtime == (vm::reservation_acquire(raddr, 128) & -128) && cmp_rdata(rdata, data))
+			else if (auto& data = vm::_ref<decltype(rdata)>(addr); rtime == (vm::reservation_acquire(raddr, 128) & -128))
 			{
-				auto& res = vm::reservation_lock(raddr, 128);
-				const u64 old_time = res.load() & -128;
-
-				if (rtime == old_time)
+				if (cmp_rdata(rdata, to_write))
 				{
-					*reinterpret_cast<atomic_t<u32>*>(&data) += 0;
-
-					// Full lock (heavyweight)
-					// TODO: vm::check_addr
-					vm::writer_lock lock(addr);
-
-					if (cmp_rdata(rdata, data))
-					{
-						mov_rdata(data, to_write);
-						res.release(old_time + 128);
-						result = 1;
-					}
-					else
-					{
-						res.release(old_time);
-					}
+					// No changes, just unlink data
+					result = 1;
```

Comment:
```suggestion
					result = cmp_rdata(rdata, data);
```

Reply:
You must advance rtime in case of succes!

Reply:
Could you explain this shortly and shall I use reservation_update() for that?

Reply:
On success (rtime == vm::reservation_acquire(raddr, 128) && cm_rdata(rdata, data) == true), the reservation value must be incremented by 128.
`result` must be set accordingly to failure or success, even if to_write didn't change you still need to check rdata == data.
You don't need to use reservation_update
use `result = cmp_rdata(rdata, data) && vm::reservation_acquire(raddr, 128).compare_and_swap_test(rtime, rtime + 128);`

Reply:
I see.

Regarding cmp_rdata(rdata,data): Do we really need this? I'm getting headaches over this. 

rdata is volatile and not read under a lock. My assumption is: No other SPU thread can have changed the data. They would have used reservation primitives and increased rtime. So only PPU could have changed it. 

In case it did not change: Everything is fine.

In case data changed: If SPU had finished earlier no one had noticed the "no-op". But now we are a little late. Why not just do as we finished the "no-op" right before the PPU?

Reply:
It's more than that, think of it as also a "feature" to detect any data changes that occur between GETALLR and PUTLLC, which is what happens in some games.
It's also called LR polling, when you constantly read from getllar and write back until there's a data change.

Reply:
It also has accuracy reasons, reservation could be lost by a local PUT write from the same SPU for example.

Reply:
Thanks for the clarification.